File: src/invoiceops_agent/api/schemas/invoice_read.py
"""Bounded invoice queue and aggregate detail response contracts."""

from typing import Literal
from uuid import UUID

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field, JsonValue

from invoiceops_agent.schemas.common import ExactDecimal
from invoiceops_agent.schemas.documents import DocumentType

InvoiceStatus = Literal[
    "RECEIVED",
    "QUEUED",
    "PROCESSING",
    "NEEDS_REVIEW",
    "APPROVED",
    "REJECTED",
    "RETURNED",
    "ARCHIVED",
    "FAILED",
]
RunStatus = Literal["QUEUED", "RUNNING", "PAUSED", "COMPLETED", "FAILED", "CANCELLED"]
ExceptionStatus = Literal["OPEN", "IN_REVIEW", "RESOLVED", "ESCALATED"]
InvoiceSource = Literal["UPLOAD", "EMAIL"]


class InvoiceReadModel(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, allow_inf_nan=False)


class InvoiceListQuery(InvoiceReadModel):
    status: InvoiceStatus | None = None
    run_status: RunStatus | None = None
    source: InvoiceSource | None = None
    exception_only: bool = False
    min_priority: int | None = Field(default=None, ge=0, le=3)
    limit: int = Field(default=50, ge=1, le=100)
    cursor: str | None = Field(default=None, max_length=200)


class InvoiceException(InvoiceReadModel):
    id: UUID
    exception_type: str = Field(min_length=1, max_length=128)
    status: ExceptionStatus
    priority: int = Field(ge=0, le=3)
    sla_due_at: AwareDatetime
    assigned_to: str | None = Field(default=None, max_length=128)
    evidence: dict[str, JsonValue]
    recommendation: dict[str, JsonValue]
    created_at: AwareDatetime


class InvoiceSummary(InvoiceReadModel):
    id: UUID
    run_id: UUID
    status: InvoiceStatus
    run_status: RunStatus
    source: InvoiceSource
    content_type: DocumentType
    vendor_name: str | None = Field(default=None, max_length=256)
    invoice_number: str | None = Field(default=None, max_length=128)
    po_number: str | None = Field(default=None, max_length=128)
    currency: str | None = Field(default=None, pattern=r"^[A-Z]{3}$")
    total_amount: ExactDecimal | None = None
    exception_id: UUID | None = None
    exception_type: str | None = Field(default=None, max_length=128)
    exception_priority: int | None = Field(default=None, ge=0, le=3)
    exception_sla_due_at: AwareDatetime | None = None
    created_at: AwareDatetime


class InvoicePage(InvoiceReadModel):
    items: list[InvoiceSummary] = Field(max_length=100)
    next_cursor: str | None = Field(default=None, max_length=200)


class InvoiceDetail(InvoiceReadModel):
    invoice: InvoiceSummary
    exception: InvoiceException | None
    evidence: dict[str, dict[str, JsonValue]]
    read_at: AwareDatetime

File: frontend/src/schemas/invoice_read.ts
import { z } from "zod";
import { jsonValueSchema } from "./ledger";

export const invoiceStatusSchema = z.enum([
  "RECEIVED", "QUEUED", "PROCESSING", "NEEDS_REVIEW", "APPROVED",
  "REJECTED", "RETURNED", "ARCHIVED", "FAILED",
]);
export const runStatusSchema = z.enum([
  "QUEUED", "RUNNING", "PAUSED", "COMPLETED", "FAILED", "CANCELLED",
]);
export const exceptionStatusSchema = z.enum(["OPEN", "IN_REVIEW", "RESOLVED", "ESCALATED"]);
export const invoiceSourceSchema = z.enum(["UPLOAD", "EMAIL"]);
const timestamp = z.string().datetime({ offset: true });
const jsonObject = z.record(z.string(), jsonValueSchema);

export const invoiceExceptionSchema = z.object({
  id: z.string().uuid(),
  exception_type: z.string().min(1).max(128),
  status: exceptionStatusSchema,
  priority: z.number().int().min(0).max(3),
  sla_due_at: timestamp,
  assigned_to: z.string().max(128).nullable(),
  evidence: jsonObject,
  recommendation: jsonObject,
  created_at: timestamp,
}).strict();

export const invoiceSummarySchema = z.object({
  id: z.string().uuid(),
  run_id: z.string().uuid(),
  status: invoiceStatusSchema,
  run_status: runStatusSchema,
  source: invoiceSourceSchema,
  content_type: z.enum(["application/pdf", "image/png", "image/jpeg"]),
  vendor_name: z.string().max(256).nullable(),
  invoice_number: z.string().max(128).nullable(),
  po_number: z.string().max(128).nullable(),
  currency: z.string().regex(/^[A-Z]{3}$/).nullable(),
  total_amount: z.string().regex(/^-?(?:0|[1-9]\d*)(?:\.\d+)?$/).nullable(),
  exception_id: z.string().uuid().nullable(),
  exception_type: z.string().max(128).nullable(),
  exception_priority: z.number().int().min(0).max(3).nullable(),
  exception_sla_due_at: timestamp.nullable(),
  created_at: timestamp,
}).strict();

export const invoicePageSchema = z.object({
  items: z.array(invoiceSummarySchema).max(100),
  next_cursor: z.string().max(200).nullable(),
}).strict();

export const pendingProposalSchema = z.object({
  id: z.string().uuid(),
  action: z.enum(["APPROVE", "RETURN", "ESCALATE"]),
  rationale: z.string(),
  reason_code: z.string(),
  actor_id: z.string(),
  created_at: timestamp,
}).strict();

export const invoiceDetailSchema = z.object({
  invoice: invoiceSummarySchema,
  exception: invoiceExceptionSchema.nullable(),
  pending_proposal: pendingProposalSchema.nullable(),
  evidence: z.record(z.string(), jsonObject),
  read_at: timestamp,
}).strict();

export type InvoiceSummary = z.infer<typeof invoiceSummarySchema>;
export type InvoicePage = z.infer<typeof invoicePageSchema>;
export type InvoiceDetail = z.infer<typeof invoiceDetailSchema>;
